### PR TITLE
Public sum/avg

### PR DESCRIPTION
**Pull request by cherylEnkidu**

### CODE REVIEW

File: firebase-firestore/api.txt [hunk 1]
```diff
@@ -19,13 +19,39 @@ package com.google.firebase {
...
   public class AggregateQuery {
     method @NonNull public com.google.android.gms.tasks.Task<com.google.firebase.firestore.AggregateQuerySnapshot> get(@NonNull com.google.firebase.firestore.AggregateSource);
+    method @NonNull public java.util.List<com.google.firebase.firestore.AggregateField> getAggregateFields();
     method @NonNull public com.google.firebase.firestore.Query getQuery();
   }
 
   public class AggregateQuerySnapshot {
+    method @Nullable public Object get(@NonNull com.google.firebase.firestore.AggregateField);
+    method @RestrictTo(androidx.annotation.RestrictTo.Scope.LIBRARY) public long get(@NonNull com.google.firebase.firestore.AggregateField.CountAggregateField);
+    method @Nullable @RestrictTo(androidx.annotation.RestrictTo.Scope.LIBRARY) public Double get(@NonNull com.google.firebase.firestore.AggregateField.AverageAggregateField);
```
Comment (MarkDuckworth) [inline firebase-firestore/api.txt:51]: Should these two API members still have the annotation `@RestrictTo(androidx.annotation.RestrictTo.Scope.LIBRARY)`?

File: firebase-firestore/api.txt [hunk 2]
```diff
@@ -19,13 +19,39 @@ package com.google.firebase {
...
+
+  public static class AggregateField.CountAggregateField extends com.google.firebase.firestore.AggregateField {
+  }
+
+  public static class AggregateField.SumAggregateField extends com.google.firebase.firestore.AggregateField {
+  }
+
   public class AggregateQuery {
     method @NonNull public com.google.android.gms.tasks.Task<com.google.firebase.firestore.AggregateQuerySnapshot> get(@NonNull com.google.firebase.firestore.AggregateSource);
+    method @NonNull public java.util.List<com.google.firebase.firestore.AggregateField> getAggregateFields();
```
Comment (MarkDuckworth) [inline firebase-firestore/api.txt:44]: We didn't include this in the public API proposal, so we should consider adding `@RestrictTo(androidx.annotation.RestrictTo.Scope.LIBRARY)`

**Review by ehsannas — CHANGES_REQUESTED** [review]
javadoc is missing from AggregateField.java

File: firebase-firestore/src/main/java/com/google/firebase/firestore/AggregateQuerySnapshot.java [hunk 3]
```diff
@@ -108,8 +103,6 @@ public long get(@Nonnull AggregateField.CountAggregateField countAggregateField)
    * @param averageAggregateField The average aggregation for which the value is requested.
    * @return The result of the given average aggregation.
    */
-  // TODO(sumavg): Remove the `hide` and scope annotations.
-  /** @hide */
   @RestrictTo(RestrictTo.Scope.LIBRARY)
```
Comment (MarkDuckworth) [inline firebase-firestore/src/main/java/com/google/firebase/firestore/AggregateQuerySnapshot.java:106]: I think this RestrictTo should be removed. This member should be in the public API.

File: firebase-firestore/src/main/java/com/google/firebase/firestore/AggregateQuerySnapshot.java [hunk 4]
```diff
@@ -87,8 +84,6 @@ public Object get(@Nonnull AggregateField aggregateField) {
    * @param countAggregateField The count aggregation for which the value is requested.
    * @return The result of the given count aggregation.
    */
-  // TODO(sumavg): Remove the `hide` and scope annotations.
-  /** @hide */
   @RestrictTo(RestrictTo.Scope.LIBRARY)
```
Comment (MarkDuckworth) [inline firebase-firestore/src/main/java/com/google/firebase/firestore/AggregateQuerySnapshot.java:87]: I think this RestrictTo should be removed. This member should be in the public API.

File: firebase-firestore/src/main/java/com/google/firebase/firestore/core/Query.java [hunk 5]
```diff
@@ -57,11 +57,17 @@ public static Query atPath(ResourcePath path) {
...
-  private List<OrderBy> memoizedOrderBy;
+  private List<OrderBy> memoizedNormalizedOrderBys;
 
-  // The corresponding Target of this Query instance.
+  /** Returns a `Target` instance this query will be mapped to in backend and local store. */
   private @Nullable Target memoizedTarget;
 
+  /**
+   * Returns a `Target` instance this query will be mapped to in backend and local store, for use
+   * within an aggregate query.
```
Comment (MarkDuckworth) [inline firebase-firestore/src/main/java/com/google/firebase/firestore/core/Query.java:67]: During Web code review, there was feedback to add additional comments describing the difference between aggregate and non-aggregate target. Consider adding [this additional info](https://github.com/firebase/firebase-js-sdk/pull/7507/files#diff-88195d5b635ef88c28a2fe6e379f36fdbbe7059dab7b92e49940da39a6335c61R77) to the comment.

**Review by MarkDuckworth — COMMENTED** [review]
I have a few comments that should be addressed, but otherwise LGTM.

File: firebase-firestore/api.txt (hunk 1 above)
Comment (cherylEnkidu) [reply firebase-firestore/api.txt:51]: Can you tell me which of these APIs need to be restricted? I checked the API proposal and I am confused :(

File: firebase-firestore/src/main/java/com/google/firebase/firestore/AggregateField.java [hunk 6]
```diff
@@ -86,43 +88,109 @@ public int hashCode() {
...
+   */
   @NonNull
   public static SumAggregateField sum(@NonNull FieldPath fieldPath) {
     return new SumAggregateField(fieldPath);
   }
 
+  /**
+   * Create an {@link AverageAggregateField} object that can be used to compute the average of a specified field over a range of documents in the result set of a query.
+   *
+   * <p>The result of an average operation will always be a 64-bit integer value, a double, or NaN.
```
Comment (MarkDuckworth) [inline firebase-firestore/src/main/java/com/google/firebase/firestore/AggregateField.java:146]: This should be updated to:
```
   * <p>The result of an average operation will always be a double or NaN.
```

I need to fix this in iOS as well.

File: firebase-firestore/src/main/java/com/google/firebase/firestore/AggregateField.java [hunk 7]
```diff
@@ -86,43 +88,109 @@ public int hashCode() {
...
+   */
   @NonNull
   public static AverageAggregateField average(@NonNull String field) {
     return new AverageAggregateField(FieldPath.fromDotSeparatedPath(field));
   }
 
+  /**
+   * Create an {@link AverageAggregateField} object that can be used to compute the average of a specified field over a range of documents in the result set of a query.
+   *
+   * <p>The result of an average operation will always be a 64-bit integer value, a double, or NaN.
```
Comment (MarkDuckworth) [inline firebase-firestore/src/main/java/com/google/firebase/firestore/AggregateField.java:164]: This should be updated to:
```
   * <p>The result of an average operation will always be a double or NaN.
```

I need to fix this in iOS as well.

**Review by MarkDuckworth — APPROVED** [review]
LGTM. We do two minor documentation tweaks. They will also be fixed in iOS.

File: firebase-firestore/src/main/java/com/google/firebase/firestore/AggregateField.java (hunk 6 above)
Comment (ehsannas) [reply firebase-firestore/src/main/java/com/google/firebase/firestore/AggregateField.java:146]: >.< I also need to fix it for Java

File: firebase-firestore/src/main/java/com/google/firebase/firestore/AggregateField.java (hunk 7 above)
Comment (ehsannas) [reply firebase-firestore/src/main/java/com/google/firebase/firestore/AggregateField.java:164]: >.< I also need to fix it for Java

Comment (MarkDuckworth) [reply firebase-firestore/src/main/java/com/google/firebase/firestore/AggregateField.java]: I fixed it for Android and ran the api doc generation and formatting tasks that were causing @cherylEnkidu trouble.

File: firebase-firestore/src/main/java/com/google/firebase/firestore/AggregateField.java (hunk 6 above)
Comment (MarkDuckworth) [reply firebase-firestore/src/main/java/com/google/firebase/firestore/AggregateField.java]: I fixed it for Android and ran the api doc generation and formatting tasks that were causing @cherylEnkidu trouble.

**Review by markarndt — APPROVED** [review]

**Review by ehsannas — APPROVED** [review]

**Review by MarkDuckworth — APPROVED** [review]

**Review by ehsannas — APPROVED** [review]

**Comment by ChinoJones** [comment]
cheryllin/sum-avg-publicconst client = require("firebase-tools");
client.apps
  .list("ANDROID", { project: "foo" })
  .then((data) => {
    // ...
  })
  .catch((err) => {
    // ...
  });

**Pull request merged**